[TF FE] Add developer documentation for TensorFlow frontend (#13505)
openvinotoolkit/openvino: src/frontends/tensorflow/README.md

# OpenVINO TensorFlow Frontend

The TensorFlow Frontend (TF FE) is a C++ based OpenVINO Frontend component that is responsible for reading and converting a TensorFlow model to an `ov::Model` object

> that further can be serialized into the Intermediate Representation (IR) format.

Reviewer: Please also add "This is an internal API for OpenVINO that is used to implement user facing API like Model Optimizer, read_model function and OpenVINO TensorFlow integration. Regular users shouldn't use the frontend directly."

Author: Added. Thanks!

TensorFlow Frontend uses [the common coding style rules](../../../docs/dev/coding_style.md).

## Key contacts

People from the [openvino-tf-frontend-maintainers](https://github.com/orgs/openvinotoolkit/teams/openvino-tf-frontend-maintainers) have the rights to approve and merge PRs to the TensorFlow Frontend component.
They can assist with any questions about the component.

## Components

The structure of OpenVINO TensorFlow Frontend sources includes the following directories:

* [include](./include) is a public frontend API.
* [src](./src/) folder contains the sources of the component.
* [unit tests](./tests) cover the internal transformations.

## Architecture

OpenVINO TensorFlow Frontend uses [TensorFlow Protobuf files](./src/proto) for reading and parsing TensorFlow model formats.

The conversion of the TensorFlow operation requires either one pass using [Loaders](./src/op) or two transformation pass
using [Loaders](./src/op) and [Internal Transformation](./src/helper_transforms).
It is often sufficient to use only [Loaders](./src/op). Two transformation pass is used when some TensorFlow operation cannot be mapped into a sub-graph of OpenVINO opset

> and its conversion depends on the succeeding operations in the graph.

Author: It is about a pass with two transformations. It looks correct to me.

# OpenVINO TensorFlow Frontend

> The TensorFlow Frontend (TF FE) is a C++ based OpenVINO Frontend component that is responsible for reading and converting TensorFlow models to the Intermediate Representation (IR) format.

Reviewer 2: ```suggestion
The TensorFlow Frontend (TF FE) is a C++ based OpenVINO Frontend component that is responsible for reading and converting TensorFlow models to the `ov::Model` representation.
```

```mermaid
flowchart BT

>     model[(model.pb)]

Reviewer 2: ```suggestion
    model[(model_1.pb)]
```

    style model fill:#427cb0

>     model2[(model.pb)]

Reviewer 2: ```suggestion
    model2[(model_2.pb)]
```

# OpenVINO TensorFlow Frontend

The TensorFlow Frontend (TF FE) is a C++ based OpenVINO Frontend component that is responsible for reading and converting TensorFlow models to the Intermediate Representation (IR) format.

```mermaid
flowchart BT
    model[(model.pb)]
    style model fill:#427cb0
    model2[(model.pb)]

>     style model2 fill:#427cb0

Author: See my comment above.

>     style model2 fill:#427cb0

Author: Fixed.

    ovtf(OpenVINO Integration with TensorFlow)

>     style ovtf fill:#ffffc2

Reviewer 2: click ovtf "https://github.com/openvinotoolkit/openvino_tensorflow"

Author: I will add it.

Author: Done.

    tf_fe(TensorFlow Frontend)
    style tf_fe fill:#ee9a4d
    fem(Frontend Manager)
    mo(Model Optimizer)
    model --> ovtf --> tf_fe

>     model2 --> mo --> fem --> tf_fe

Author: I would like to have two separate models to show that it can be passed either to MO or to OVITF

>     model2 --> mo --> fem --> tf_fe

Reviewer 2: In this case I propose to use different names

    model2 --> mo --> fem --> tf_fe
    click ovtf "https://github.com/openvinotoolkit/openvino_tensorflow"
```

> Currently, it is only used by [OpenVINO Integration with TensorFlow Frontend](https://github.com/openvinotoolkit/openvino_tensorflow).

Reviewer: ```suggestion
Currently, it is only used by [OpenVINO Integration with TensorFlow](https://github.com/openvinotoolkit/openvino_tensorflow).
```

> Model Optimizer for now relies on the legacy TensorFlow Frontend developed in Python.

Reviewer: ```suggestion
Model Optimizer for now relies on the legacy [TensorFlow Frontend] (https://docs.openvino.ai/latest/openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow.html) developed in Python.
```

> TensorFlow Frontend uses [the common coding style rules](../../../docs/dev/coding_style.md).

Reviewer: Doesn't seem to be worth mentioning, because there is nothing special with TF FE or any other code inside OpenVINO that follows these rules.

Author: Agree that this is not important or looks like much off-top:)

People from the [openvino-tf-frontend-maintainers](https://github.com/orgs/openvinotoolkit/teams/openvino-tf-frontend-maintainers) have the rights to approve and merge PRs to the TensorFlow Frontend component.
They can assist with any questions about the component.

## Components

The structure of OpenVINO TensorFlow Frontend sources includes the following directories:

* [include](./include) is a public frontend API.
* [src](./src/) folder contains the sources of the component.

> * [unit tests](./tests) cover the internal transformations.

Reviewer 2: ```suggestion
* [tests](./tests) cover the internal transformations.
```

* [src](./src/) folder contains the sources of the component.
* [tests](./tests) cover the internal transformations.

## Architecture

OpenVINO TensorFlow Frontend uses [TensorFlow Protobuf files](./src/proto) for reading and parsing TensorFlow model formats.

The conversion of the TensorFlow operation requires either one pass using [Loaders](./src/op) or two transformation passes

> using [Loaders](./src/op) and [Internal Transformation](./src/helper_transforms).

Reviewer: It is not clear what does it mean without reading that part about architecture. It is better to move the complete architecture article here. It is not too big to be separated to a different page.

Author: Agree, let us do it.

Author: Done

> It is often sufficient to use only [Loaders](./src/op). Two transformation pass is used when some TensorFlow operation cannot be mapped into a sub-graph of OpenVINO opset

Reviewer 2: Three times we use `Loaders` here. Can we reduce it?

Author: Done.
